betterproto/plugin.py

```diff
@@ -174,266 +212,253 @@ def generate_code(request, response):
     )
     template = env.get_template("template.py")
 
-    output_map = {}
+    output_files = collections.defaultdict()
+
     for proto_file in request.proto_file:
-        out = proto_file.package
-        if out == "google.protobuf":
+        package = proto_file.package
+        if package == "google.protobuf":
             continue
 
-        if not out:
-            out = os.path.splitext(proto_file.name)[0].replace(os.path.sep, ".")
-
-        if out not in output_map:
-            output_map[out] = {"package": proto_file.package, "files": []}
-        output_map[out]["files"].append(proto_file)
+        if package:
+            output_file_name = os.path.sep.join(package.split('.') + ['__init__.py'])
+        else:
+            output_file_name = '__root__.py'
 
-    # TODO: Figure out how to handle gRPC request/response messages and add
-    # processing below for Service.
+        output_files.setdefault(output_file_name, {"package": package, "files": []})
+        output_files[output_file_name]["files"].append(proto_file)
 
-    for filename, options in output_map.items():
-        package = options["package"]
-        # print(package, filename, file=sys.stderr)
-        output = {
+    for output_file_name, output_file_data in output_files.items():
+        package = output_file_data["package"]
+        template_data = {
             "package": package,
-            "files": [f.name for f in options["files"]],
+            "files": [f.name for f in output_file_data["files"]],
             "imports": set(),
             "datetime_imports": set(),
             "typing_imports": set(),
             "messages": [],
             "enums": [],
             "services": [],
         }
+        output_file_data['template_data'] = template_data
 
-        type_mapping = {}
-
-        for proto_file in options["files"]:
-            # print(proto_file.message_type, file=sys.stderr)
-            # print(proto_file.service, file=sys.stderr)
-            # print(proto_file.source_code_info, file=sys.stderr)
-
+    # read messages
+    global_messages = []
+    for output_file_name, output_file_data in output_files.items():
+        for proto_file in output_file_data["files"]:
```

Comment:
this bit could be streamlined a bit more since you're here:

```python
global_messages = []
for _, output_file_data in output_files.items():
    output_file_data['template_data'] =  {
        "package": output_file_data["package"],
        "files": [f.name for f in output_file_data["files"]],
        "imports": set(),
        "datetime_imports": set(),
        "typing_imports": set(),
        "messages": [],
        "enums": [],
        "services": [],
    }

    # read messages
    for proto_file in output_file_data["files"]:
        for item, path in traverse(proto_file):
            read_item(
                item, 
                path, 
                proto_file,
                # < No need to pass packages seperately becuase it's in template_data >
                output_file_data['template_data'], 
                global_messages
            )
```